kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,624 @@
+type TableChangesScanData = (Box<dyn EngineData>, Vec<bool>, Arc<HashMap<String, DvInfo>>);
+
+/// Given an iterator of ParsedLogPath and a predicate, returns an iterator of
+/// [`TableChangesScanData`] = `(engine_data, selection_vec, remove_deletion_vectors)`.
+/// Each row that is selected in the returned `engine_data` _must_ be processed to
+/// complete the scan. Non-selected rows _must_ be ignored.
+pub(crate) fn table_changes_action_iter(
+    engine: &dyn Engine,
+    commit_files: impl IntoIterator<Item = ParsedLogPath>,
+    table_schema: &SchemaRef,
+    predicate: Option<ExpressionRef>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+    let json_client = engine.get_json_handler();
+    let filter = DataSkippingFilter::new(engine, table_schema, predicate);
+
+    let expression_evaluator = engine.get_expression_handler().get_evaluator(
+        get_log_add_schema().clone(),
+        get_add_transform_expr(),
+        SCAN_ROW_DATATYPE.clone(),
+    );
+
+    let result = commit_files
+        .into_iter()
+        .map(move |commit_file| -> DeltaResult<_> {
+            let mut scanner = LogReplayScanner::new(
+                commit_file,
+                json_client.clone(),
+                filter.clone(),
+                expression_evaluator.clone(),
+            );
+            scanner.phase_1()?;
+            scanner.resolve_dvs();
+            scanner.into_scan_batches()
+        })
+        .flatten_ok()
+        .map(|res| res?);
+    Ok(result)
+}
+
+fn get_add_transform_expr() -> Expression {
+    Expression::Struct(vec![
+        column_expr!("add.path"),
+        column_expr!("add.size"),
+        column_expr!("add.modificationTime"),
+        column_expr!("add.stats"),
+        column_expr!("add.deletionVector"),
+        Expression::Struct(vec![column_expr!("add.partitionValues")]),
+    ])
+}
+
+struct LogReplayScanner {
+    has_cdc_action: bool,
+    // A map from path to the deletion vector from the remove action. It is guaranteed that there
+    // is an add action with the same path in this commit
+    remove_dvs: HashMap<String, DvInfo>,
+    commit_file: ParsedLogPath,
+    json_client: Arc<dyn JsonHandler>,
+    timestamp: i64,
+    filter: Option<DataSkippingFilter>,
+    expression_evaluator: Arc<dyn ExpressionEvaluator>,
+    add_paths: HashSet<String>,
+}
+
+impl LogReplayScanner {
+    fn new(
+        commit_file: ParsedLogPath,
+        json_client: Arc<dyn JsonHandler>,
+        filter: Option<DataSkippingFilter>,
+        expression_evaluator: Arc<dyn ExpressionEvaluator>,
+    ) -> Self {
+        Self {
+            timestamp: commit_file.location.last_modified,
+            commit_file,
+            json_client,
+            filter,
+            has_cdc_action: Default::default(),
+            remove_dvs: Default::default(),
+            expression_evaluator,
+            add_paths: Default::default(),
+        }
+    }
+    fn phase_1(&mut self) -> DeltaResult<()> {
```

Comment:
i'd have a slight preference for naming this `X_phase` where `X` describes what this phase actually does

Reply:
Yep, I'm still trying to figure out a good name. I was thinking `prepare_phase` or `prepare`